core/trino-main/src/main/java/io/trino/operator/scalar/CharacterStringCasts.java

```diff
@@ -92,68 +87,128 @@ public static Slice charToVarcharCast(@LiteralParameter("x") Long x, @LiteralPar
     @LiteralParameters({"x", "y"})
     public static Slice varcharToCharSaturatedFloorCast(@LiteralParameter("y") long y, @SqlType("varchar(x)") Slice slice)
     {
-        List<Integer> codePoints = new ArrayList<>(toCodePoints(slice));
+        IntList codePoints = toCodePoints(slice);
 
         // if Varchar(x) value length (including spaces) is greater than y, we can just truncate it
         if (codePoints.size() >= y) {
             // char(y) slice representation doesn't contain trailing spaces
-            codePoints = trimTrailing(codePoints, ' ');
-            List<Integer> codePointsTruncated = codePoints.stream()
-                    .limit(y)
-                    .collect(toImmutableList());
-            return codePointsToSliceUtf8(codePointsTruncated);
+            codePoints.truncate(Math.min((int) y, codePoints.size));
+            trimTrailing(codePoints, ' ');
+            return codePointsToSliceUtf8(codePoints);
         }
 
         /*
          * Value length is smaller than same-represented char(y) value because input varchar has length lower than y.
          * We decrement last character in input (in fact, we decrement last non-zero character) and pad the value with
          * max code point up to y characters.
          */
-        codePoints = trimTrailing(codePoints, '\0');
+        trimTrailing(codePoints, '\0');
 
-        if (codePoints.isEmpty()) {
+        if (codePoints.size() == 0) {
             // No non-zero characters in input and input is shorter than y. Input value is smaller than any char(4) casted back to varchar, so we return the smallest char(4) possible
-            return codePointsToSliceUtf8(nCopies(toIntExact(y), (int) '\0'));
+            return Slices.allocate((int) y);
         }
 
-        codePoints = new ArrayList<>(codePoints);
         codePoints.set(codePoints.size() - 1, codePoints.get(codePoints.size() - 1) - 1);
-        codePoints.addAll(nCopies(toIntExact(y) - codePoints.size(), Character.MAX_CODE_POINT));
+        int toAdd = toIntExact(y) - codePoints.size();
+        for (int i = 0; i < toAdd; i++) {
+            codePoints.add(Character.MAX_CODE_POINT);
+        }
 
         verify(codePoints.get(codePoints.size() - 1) != ' '); // no trailing spaces to trim
 
         return codePointsToSliceUtf8(codePoints);
     }
 
-    private static List<Integer> trimTrailing(List<Integer> codePoints, int codePointToTrim)
+    /**
+     * This is a very primitive version of dynamically-sized int list.
+     * It provides only limited bound-checking.
+     * It should not be used anywhere outside of this class.
+     */
+    private static class IntList
```

Comment:
Are you confident that the use-case here is central enough to go with a dedicated List implementation?

@sopel39 what can we use here instead?

Reply:
We can go with static int array of maximum possible size. It will still take less memory than the current solution.
For slice of `n` ascii chars:
static array - `4*n`
dynamic array - `4*n` * ~1.5
List<Integer> - `8*n` * ~1.2
For slice of `n` bytes filed with 4-byte chars:
static array - `4*n`
dynamic array - `n * ~1.5`
List<Integer> - `2*n * ~1.2 + ~6*n`

Reply:
why not to use `it.unimi.dsi.fastutil.ints.IntArrayList`? I'm not sure bound checks are that expensive.

Reply:
Somehow I was under the impression that it is not available there. But it is, so I guess case closed